src/ipld/util.rs

```diff
@@ -194,3 +193,146 @@ fn should_save_block_to_snapshot(cid: &Cid) -> bool {
         )
     }
 }
+
+use crate::shim::clock::ChainEpoch;
+use futures::Stream;
+use fvm_ipld_blockstore::Blockstore;
+use pin_project_lite::pin_project;
+use std::pin::Pin;
+use std::task::{Context, Poll};
+
+enum Task {
+    // Yield the block, don't visit it.
+    Emit(Cid),
+    // Visit all the elements, recursively.
+    Iterate(Ipld),
+}
+
+pin_project! {
+    struct ChainStream<DB, T> {
+        #[pin]
+        tipset_iter: T,
+        db: DB,
+        dfs: VecDeque<Task>, // Depth-first work queue.
+        seen: CidHashSet,
+        stateroot_limit: ChainEpoch,
+    }
+}
+
+/// Initializes a stream of blocks.
+///
+/// # Arguments
+///
+/// * `db` - A database that implements [`Blockstore`] interface.
+/// * `tipset_iter` - An iterator of [`Tipset`], descending order `$child -> $parent`.
+/// * `stateroot_limit` - An epoch that signifies how far back we need to inspect tipsets.
+/// in-depth. This has to be pre-calculated using this formula: `$cur_epoch - $depth`, where
+/// `$depth` is the number of `[`Tipset`]` that needs inspection.
+pub fn stream_chain<DB: Blockstore, T: Iterator<Item = Tipset> + Unpin>(
+    db: DB,
+    tipset_iter: T,
+    stateroot_limit: ChainEpoch,
+) -> impl Stream<Item = anyhow::Result<(Cid, Vec<u8>)>> {
+    ChainStream {
+        tipset_iter,
+        db,
+        dfs: VecDeque::new(),
+        seen: CidHashSet::default(),
+        stateroot_limit,
+    }
+}
+
+impl<DB: Blockstore, T: Iterator<Item = Tipset> + Unpin> Stream for ChainStream<DB, T> {
+    type Item = anyhow::Result<(Cid, Vec<u8>)>;
+
+    fn poll_next(self: Pin<&mut Self>, _: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        use Task::*;
+        let mut this = self.project();
+
+        let stateroot_limit = *this.stateroot_limit;
+        loop {
+            while let Some(ipld) = this.dfs.pop_front() {
+                match ipld {
+                    Emit(cid) => {
+                        let result = this.db.get(&cid);
+
+                        let result = result.and_then(|val| {
+                            let block = val.ok_or(anyhow::anyhow!("missing key"))?;
+                            Ok((cid, block))
+                        });
+
+                        return Poll::Ready(Some(result));
+                    }
+                    Iterate(Ipld::Null) => {}
```

Comment:
can do `Iterate(Ipld::Null) | Iterate(Ipld::Bool(_))) | ... => {}`

Reply:
Pushing to follow-up PR.